Feature/update readme (#78)
fivetran/dbt_salesforce_formula_utils: README.md

</p>

# Fivetran Salesforce Formula Utils
# 📣 What does this dbt package do?

> This package includes macros to be used within a Salesforce dbt project to accurately map Salesforce Formulas to existing tables.

Reviewer: I would perhaps include some links to the salesforce ERD and connector to be in line with some of our other READMEs, for ex google_ads https://github.com/fivetran/dbt_google_ads_source#-what-does-this-dbt-package-do

In particular, you will need the following sources defined in your `src_salesforce.yml` file:
```yml
version: 2

sources:
  - name: salesforce # It would be best to keep this named salesforce
    schema: 'salesforce_schema' # Modify this to be where your Salesforce data resides
    # Add the database where your Salesforce data resides if different from the target database. Eg. 'my_salesforce_database'. By default the target.database is used.
    tables:
      - name: fivetran_formula_model
        description: Used for the recommended Option 1 version of the formula solution.
      - name: fivetran_formula
        description: Used for options 2 and 3 of the original individual formula solution.

>       ## Any other source tables you are creating models for should be defined here as well. They aren't required, but it is best organizational practice and allows Fivetran to compile data lineage graphs

Author: ah yeah i agree, especially if folks are using the salesforce dbt package

>       ## Any other source tables you are creating models for should be defined here as well. They aren't required, but it is best organizational practice and allows Fivetran to compile data lineage graphs

Author: added

```

## Step 4: Create models

> ### (Recommended and default) Option 1: Generate all relevant formula fields using connector-made query

Author: yeah kinda, what i mean here is:
- default = if the customer doesn't provide any arguments at all besides the `source_table` to the macro, this option would run (which maybe i should highlight more?) 
- recommended = this option uses a much better lookup table

> This option makes use of the `fivetran_formula_model` lookup table, which stores connector-generated SQL queries for each source table. Compared to `fivetran_formula`, which is used in Options 2 & 3, it is typically more complete and supports most double-nested formulas. 

### Option 2: Generate all relevant formula fields using package-made query

If you would like your model to generate all the formula fields related to your source table then you will need to: 
1. Create a new file in your models folder and name it `your_table_name_here`.sql (e.g. `customer.sql`; this is not necessary but recommended as best practice). 
2. Add the below snippet calling the `sfdc_formula_view` macro into the file. Update the `source_table` argument to be the source table name for which you are generating the model (e.g. `customer`).
```sql
{{ salesforce_formula_utils.sfdc_formula_view(
    source_table='your_source_table_name_here',
    full_statement_version=false) 
}}
```

**Output**: All formulas for the chosen source table will be generated in a `select` statement. This output differs from Option 1 only in the format of compiled code. 

> > This option makes use of the `fivetran_formula` lookup table, which requires the package to combine fields' formulas into a SQL query for each source table. This option does not support double-nested formulas and may be incomplete compared to Option #1.

Author: @fivetran-joemarkiewicz i'm actually not 100% sure on the answer here ^, are you?

If you would like your model to generate only a specified subset of your formula fields related to your source table then you will need to: 
1. Create a new file in your models folder and name it `your_table_name_here`.sql (e.g. `customer.sql`; this is not necessary but recommended as best practice). 
2. Add the below snippet calling the `sfdc_formula_view` macro into the file and:
    - Update the `source_table` argument to be the source table name for which you are generating the model (e.g. `customer`).
    - Update the `fields_to_include` argument to contain all the fields from your source that you would like to be included in the final output. Be sure that the field(s) you would like to include are enclosed within brackets as an array (ie. `[]`)
```sql
{{ salesforce_formula_utils.sfdc_formula_view(
    source_table='your_source_table_name_here', 
    fields_to_include=['i_want_this_field','also_this_one','maybe_a_third_as_well','lets_add_more'],
    full_statement_version=false) 
}}
```

**Output**: Formulas provided in the `fields_to_include` variable will be generated in a `select` statement for the chosen source table.

> > This option makes use of the `fivetran_formula` lookup table, which requires the package to combine fields' formulas into a SQL query for each source table. This option does not support double-nested formulas.

Reviewer: Just for consistency

```suggestion
> This option makes use of the `fivetran_formula` lookup table, which requires the package to combine fields' formulas into a SQL query for each source table. This option does not support double-nested formulas and therefore may be incomplete compared to Option #1.
```

### Automate model creation
If you have multiple models you need to create, you can also Leverage the [sfdc_formula_model_automation](https://github.com/fivetran/dbt_salesforce_formula_utils/blob/main/sfdc_formula_model_automation.sh) script within this project to automatically create models locally via the command line. Below is an example command to copy and edit.

```bash
source dbt_modules/salesforce_formula_utils/sfdc_formula_model_automation.sh "../path/to/directory" "desired_table_1,desired_table_2,desired_table_infinity"
```

> By default, these models will run Option #1.

Author: yes, they will need to add `full_statement_version=false` for both options and `fields_to_include` to option 3. i'll add that!

> By default, these models will run Option #1.

Reviewer: so they would just need to add it to the end of the end like 
```
source dbt_modules/salesforce_formula_utils/sfdc_formula_model_automation.sh "../path/to/directory" "desired_table_1,desired_table_2,desired_table_infinity" "full_statement_version=false" "fields_to_include" 
```
Or do we mean to the .sh script?

> By default, these models will run Option #1.

Author: nah like manually in each file they want it in

> By default, these models will run Option #1.

Author: which i can highlight!

> By default, these models will run Option #1.

Author: added, whatcha think

## Step 5: Exclude problematic formula fields

> The `sfdc_formula_view` macro has been created to allow for two degrees of formula field reference. For example:

Reviewer: I think two degrees makes sense, but I also noticed we referenced "double nested" a lot and I think it would make sense to use that instead here. I think the below are great examples of what a double nested formula would be and we could take advantage of this section to explain what "double nested" means for the reader.

Reviewer: Based on the above `README`..

```suggestion
The `sfdc_formula_view` macro has been created to support double-nested formula field references for most cases using Option #1, however, never if using Options #2 or #3. For example:
```

- :white_check_mark: : A formula field references standard fields from the base Salesforce table.
- :white_check_mark: : A formula field references another formula field that does **not** reference other formula fields.

> - 🚧     : A formula field references another formula field that references another formula field (and so on...). This may be possible for certain situations using Option #1 above, but never for Options 2 & 3.

Reviewer: This is kind of an anti-example and so including it in this section is a bit confusing. 

```suggestion
The `sfdc_formula_view` macro may be able to support some cases of n-nested formula field references if using Option #1 above, however, never if using Options #2 or #3. For example:
- 🚧     : A formula field references another formula field that references another formula field (and so on...). 
```

</p>

# Fivetran Salesforce Formula Utils
# 📣 What does this dbt package do?
This package includes macros and scipts to be used within a dbt project to accurately map Salesforce Formulas to existing tables. It is designed to  work with data from [Fivetran's Salesforce connector](https://fivetran.com/docs/applications/salesforce) in the format described by [this ERD](https://fivetran.com/docs/applications/salesforce#schema).

> Note: this package is distinct from the [Salesforce dbt package](https://github.com/fivetran/dbt_salesforce), which _transforms_ Salesforce data and outputs analytics-ready end models.

# 🎯 How do I use the dbt package?
## Step 1: Prerequisites
To use this dbt package, you must have the following:
- At least one Fivetran Salesforce connector syncing data into your destination. 
- A **BigQuery**, **Snowflake**, **Redshift**, or **PostgreSQL** destination.

## Step 2: Install the package

> The [`sfdc_formula_view`](https://github.com/fivetran/dbt_salesforce_formula_utils#sfdc_formula_view-source) macro is intended to be leveraged within a dbt project that uses the source tables from Fivetran's Salesforce connector. To leverage the macro, you will add the below configuration to your `packages.yml` file (if you do not have a `packages.yml` file, create on in your root dbt project).

Reviewer: ```suggestion
The [`sfdc_formula_view`](https://github.com/fivetran/dbt_salesforce_formula_utils#sfdc_formula_view-source) macro is intended to be leveraged within a dbt project that uses the source tables from Fivetran's Salesforce connector. To leverage the macro, you will add the below configuration to your `packages.yml` file (if you do not have a `packages.yml` file, create one in your root dbt project).
```

```bash
source dbt_modules/salesforce_formula_utils/sfdc_formula_model_automation.sh "../path/to/directory" "desired_table_1,desired_table_2,desired_table_infinity"
```

**Output**: Model files for each table, populated with `{{ salesforce_formula_utils.sfdc_formula_view(source_table='table_name') }}`. By default, these models will run Option #1. To use Options #2 or #3, you will need to add `full_statement_version=false`. For Option #3, you will need to add the `fields_to_include` argument(s).

> Note: In order for this command to work, you must currently be within the root directory of your dbt project. 

## Step 5: Exclude problematic formula fields
The `sfdc_formula_view` macro has been created to support double-nested formula field references for most cases using Option #1, however, never if using Options #2 or #3. For example:
- :white_check_mark: : A formula field references standard fields from the base Salesforce table.
- :white_check_mark: : A formula field references another formula field that does **not** reference other formula fields.

> The `sfdc_formula_view` macro may be able to support some cases of n-nested formula field references if using Option #1 above, however, never if using Options #2 or #3. For example:

Reviewer: Think you just need a white space before this line or else the spacing turns out a bit weird in the final readme output
```suggestion

The `sfdc_formula_view` macro may be able to support some cases of n-nested formula field references if using Option #1 above, however, never if using Options #2 or #3. For example:
```

```
**Args:**
* `source_table` (required): The table with which you are joining the formula fields.

> * `source_name` (optional, default = `'salesforce'`): The dbt source containing the table you want to join with formula fields. Must also contain the `fivetran_formula` table.

Reviewer: Just for my knowledge, can you give an example for this field to be included and declared? I'm not sure I understand what is meant by joining a table with formula fields

Author: yeah so like if a customer [declared](https://github.com/fivetran/dbt_salesforce_formula_utils#source-tables-required) the source with a different name (we recommend salesforce), such as `fivetran_salesforce`, they'd include `source_name='fivetran_salesforce'` as an argument

Author: the language is weird tho, lemme update

> * `fields_to_include` (optional, default = `none`): If a users wishes to only run the formula fields macro for designated fields then they may be applied within this variable. This variable will ensure the model only generates the sql for the designated fields. 

Reviewer: Does this only work if `full_statement_version` is set to false?

Author: yeah, lemme add that

* `full_statement_version` (optional, default = `true`): Allows a user to leverage the `fivetran_formula_table` version of the macro which will generate the formula fields via the complete sql statement, rather than individual formulas being generated within the macro.
* `using_quoted_identifiers` (optional, default = `false`): For warehouses with case sensitivity enabled this argument **must** be set to `true` in order for the underlying macros within this project to properly compile and execute successfully. 
* `materialization` (optional, default = `view`): By default the model will be materialized as a view. If you would like to materialize as a table, you can adjust using this argument.
> Note: If you populate the `fields_to_include` argument then the package will exclusively look for those fields. If you have designated a field to be excluded within the `sfdc_exclude_formulas` variable, then this will be ignored and the field will be included in the final model.
----

## sfdc_formula_model_automation.sh ([source](sfdc_formula_model_automation.sh))

> This bash script is intended to be used in order to automatically create the desired salesforce models via the command line within your dbt project. This bash script will generate a model file within your dbt project that contains the [sfdc_formula_view](https://github.com/fivetran/dbt_salesforce_formula_utils/blob/main/macros/sfdc_formula_view.sql) macro for the appropriately defined table(s). In order for this command to work you must be within the root directory of your dbt project. 

Reviewer: Mayhaps include something explicitly stating that this only works for option #1 (unless there is indeed some way to specify otherwise)
